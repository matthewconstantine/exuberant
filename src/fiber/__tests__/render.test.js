import React from 'react'
import fs from 'fs-extra'
import { renderElement, rerenderElement, Dir, File } from '../'

const renderChange = (oldElement, newElement) =>
  rerenderElement(newElement, renderElement(oldElement, 'output'))

const Text = ({ children }) => (children.join ? children.join('\n') : children)

describe('Render', () => {
  beforeEach(() => fs.resetMock())

  describe('Dir', () => {
    it('Creates itself', () => {
      const element = <Dir name="shasta" />
      renderElement(element, 'output')
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Renames itself', () => {
      const first = <Dir name="shasta" />
      const second = <Dir name="faye" />
      renderChange(first, second)
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Deletes itself', () => {
      const first = (
        <Dir name="wrapper">
          <Dir name="deleteMe" />
        </Dir>
      )
      const second = <Dir name="wrapper" />
      renderChange(first, second)
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Throws when no name is provided', () => {
      const element = <Dir />
      const shouldError = () => {
        renderElement(element, 'output')
      }
      expect(shouldError).toThrowErrorMatchingSnapshot()
    })
  })

  describe('File', () => {
    it('Creates a file', () => {
      const element = <File name="Sportello" />
      renderElement(element, 'output')
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Renames a file', () => {
      const first = <File name="Doc" />
      const second = <File name="Sportello" />
      renderChange(first, second)
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Deletes a file', () => {
      const first = (
        <Dir name="wrapper">
          <File name="deleteMe" />
        </Dir>
      )
      const second = <Dir name="wrapper" />
      renderChange(first, second)
      expect(fs.snapshot()).toMatchSnapshot()
    })

    it('Throws when no name is provided', () => {
      const element = <File />
      const shouldError = () => {
        renderElement(element, 'output')
      }
      expect(shouldError).toThrowErrorMatchingSnapshot()
    })

    it('Renders plain text', () => {
      const element = <File name="Sportello">Ahhh!</File>
      renderElement(element, 'output')
      expect(fs.getPath('output/Sportello')).toMatchSnapshot()
    })

    it('Mutates plain text', () => {
      const first = <File name="Sportello">Ahhh!</File>
      const second = <File name="Sportello">Whats that?</File>
      renderChange(first, second)
      expect(fs.getPath('output/Sportello')).toMatchSnapshot()
    })

    it('Deletes plain text', () => {
      const first = <File name="Sportello">Ahhh!</File>
      const second = <File name="Sportello" />
      renderChange(first, second)
      expect(fs.getPath('output/Sportello')).toMatchSnapshot()
    })

    it('Deletes multi-child text', () => {
      const first = (
        <File name="Sportello">
          <Text>Text</Text>
          <Text>deleteMe</Text>
        </File>
      )
      const second = (
        <File name="Sportello">
          <Text>Text</Text>
        </File>
      )
      renderChange(first, second)
      expect(fs.getPath('output/Sportello')).toMatchSnapshot()
    })

    it('Reorders multi-child text', () => {
      const first = (
        <File name="Sportello">
          <Text>First</Text>
          <Text>Second</Text>
          <Text>Third</Text>
        </File>
      )
      const second = (
        <File name="Sportello">
          <Text>First</Text>
          <Text>Third</Text>
          <Text>Second</Text>
        </File>
      )
      renderChange(first, second)
      expect(fs.getPath('output/Sportello')).toMatchSnapshot()
    })
  })
})
