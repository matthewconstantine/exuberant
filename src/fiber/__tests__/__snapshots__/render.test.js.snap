// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Dir Creates itself 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "ensureDirSync",
      "path": "output/shasta",
    },
  ],
  "volume": Object {
    "output": true,
    "output/shasta": true,
  },
}
`;

exports[`Render Dir Deletes itself 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "ensureDirSync",
      "path": "output/wrapper",
    },
    Object {
      "action": "ensureDirSync",
      "path": "output/wrapper/deleteMe",
    },
    Object {
      "action": "removeSync",
      "path": "output/wrapper/deleteMe",
    },
  ],
  "volume": Object {
    "output": true,
    "output/wrapper": true,
  },
}
`;

exports[`Render Dir Renames itself 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "ensureDirSync",
      "path": "output/shasta",
    },
    Object {
      "action": "renameSync",
      "newPath": "output/faye",
      "oldPath": "output/shasta",
    },
  ],
  "volume": Object {
    "output": true,
    "output/faye": true,
  },
}
`;

exports[`Render Dir Throws when no name is provided 1`] = `"Dir does not have props.name"`;

exports[`Render File Creates a file 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
  ],
  "volume": Object {
    "output": true,
    "output/Sportello": "",
  },
}
`;

exports[`Render File Deletes a file 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "ensureDirSync",
      "path": "output/wrapper",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
    Object {
      "action": "removeSync",
      "path": "output/wrapper/deleteMe",
    },
  ],
  "volume": Object {
    "output": true,
    "output/wrapper": true,
  },
}
`;

exports[`Render File Deletes multi-child text 1`] = `"Text"`;

exports[`Render File Deletes plain text 1`] = `""`;

exports[`Render File Mutates plain text 1`] = `"Whats that?"`;

exports[`Render File Renames a file 1`] = `
Object {
  "ledger": Array [
    Object {
      "action": "ensureDirSync",
      "path": "output",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
    Object {
      "action": "renameSync",
      "newPath": "output/Sportello",
      "oldPath": "output/Doc",
    },
    Object {
      "action": "writeFileSync",
      "contents": "",
    },
  ],
  "volume": Object {
    "output": true,
    "output/Sportello": "",
  },
}
`;

exports[`Render File Renders plain text 1`] = `"Ahhh!"`;

exports[`Render File Reorders multi-child text 1`] = `
"First
Third
Second"
`;

exports[`Render File Throws when no name is provided 1`] = `"File node was not provided props.name"`;
